### ISSUE TITLE
Does Gradient Accumulation work with ghost clipping?

### ISSUE THREAD

**Issue by RobRomijnders** [pytorch/opacus#683]
# Gradient accumulation and Ghost clipping

The DP optimizer for Ghost clipping enforces accumulated iterations = 1 [here](https://github.com/pytorch/opacus/blob/main/opacus/optimizers/optimizer_fast_gradient_clipping.py#L106).
The gradients, after clipping and noising, are scaled [here](https://github.com/pytorch/opacus/blob/main/opacus/optimizers/optimizer.py#L488).

However, for ghost-clipping, i.e. when `DPOptimizerFastGradientClipping` is used, in combination with gradient accumulation, this will only divide by `expected_batch_size`. Therefore, my question regards documentation:

When the user uses gradient accumulation, are they expected to overwrite `DPOptimizerFastGradientClipping.accumulated_iterations` themselves?

Thank you in advance,
Rob

**Comment by EnayatUllah** [comment]
You are right -- gradient accumulation is currently not supported with Ghost Clipping. However, you can use [Batch Memory Manager](https://opacus.ai/api/batch_memory_manager.html), to accumulate gradients over virtual mini-batches, which is supported with Ghost Clipping. Let us know if it addresses your requirements.

**Comment by EnayatUllah** [comment]
We plan to support it soon - if you want to contribute a PR towards it, we would be happy to help.

And yes, ghost clipping is supported with DPDDP -- essentially the optimizer needs to be changed to [DistributedDPOptimizerFastGradientClipping](https://github.com/pytorch/opacus/blob/main/opacus/optimizers/ddpoptimizer_fast_gradient_clipping.py).
